File: 백준/Gold/15684. 사다리 조작/사다리 조작.py
import sys

n,m,h = map(int,sys.stdin.readline().split())

# 2차원 배열로 사다리 만들기
ladder = [[0] * n for _ in range(h)]

for _ in range(m):
  a,b = map(int,sys.stdin.readline().split())
  ladder[a-1][b-1] = 1

# i번 세로선의 결과가 i번이 나오는지 확인하는 check 함수
def check():
  for i in range(n):
    tmp = i
    for j in range(h):
      if ladder[j][tmp] == 1:
        tmp += 1
      elif tmp > 0 and ladder[j][tmp-1] == 1:
        tmp -= 1
    if tmp != i:
      return False
  return True

# 사다리 만드는 DFS 함수
def dfs(cnt, x, y):
    global ans
    if ans <= cnt:
        return
    if check():
        ans = min(ans, cnt)
        return
    if cnt == 3:
        return
    for i in range(x, h):
        if i == x:
          tmp = y 
        else:
          tmp = 0
        for j in range(tmp, n - 1):
            if ladder[i][j]:
                j += 1
            else:
                ladder[i][j] = 1
                dfs(cnt + 1, i, j + 2)
                ladder[i][j] = 0

ans = 4
dfs(0, 0, 0)
print(ans if ans < 4 else -1)

File: 백준/Gold/15684. 사다리 조작/사다리 조작.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
	static int[][] rladders;

	static boolean check(int[][] ladders) {
		for (int i = 1; i < N + 1; i++) {
			int x = 0;
			int y = i;
			while (x != H + 1) {
				if (ladders[x][y] != 0) {
					y = ladders[x][y];
				}
				x++;
			}
			if (y != i)
				return false;
		}
		return true;
	}

	static void dfs(int depth, int[][] ladders, int x) {
		if (depth > 3 || depth >= ans) {
			return;
		}
		if (check(ladders)) {
			ans = Math.min(depth, ans);
			return;
		}
		for (int i = x; i < H + 1; i++) {
			for (int j = 1; j < N; j++) {
				if (ladders[i][j] == 0 && ladders[i][j + 1] == 0) {
					ladders[i][j] = j + 1;
					ladders[i][j + 1] = j;
					dfs(depth + 1, ladders, i );
					ladders[i][j] = 0;
					ladders[i][j + 1] = 0;
				}
			}
		}
	}

	static int N, M, H, ans;

	public static void main(String[] args) throws IOException {
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		H = Integer.parseInt(st.nextToken());
		ans = Integer.MAX_VALUE;
		rladders = new int[H + 1][N + 1];
		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			rladders[a][b] = b + 1;
			rladders[a][b + 1] = b;
		}
		dfs(0, rladders, 1);
		if(ans > 3) ans = -1;
		bw.write(ans + "");
		bw.close();
	}

}